docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -739,121 +779,197 @@ Specifically, `index` is in such a format:
-
-To achieve this goal, the placement rules should be marked with the source  where they come from.
+We may want to ensure that users that have users in the EU store their data in a specific location. On the surface this looks straight forward:
 
-### Building placement rules
+```sql
+CREATE TABLE users (
+	id INT NOT NULL auto_increment,
+	username VARCHAR(64) NOT NULL,
+	email VARCHAR(64) NOT NULL,
+	dateofbirth DATE NOT NULL,
+	country VARCHAR(10) NOT NULL,
+	PRIMARY KEY (id),
+	UNIQUE (username)
+) PARTITION BY LIST COLUMNS (country) (
+	PARTITION pEurope VALUES IN ('DE', 'FR', 'GB') PLACEMENT POLICY='europe',
+	PARTITION pOther VALUES IN ('US', 'CA', 'MX')
+);
+```
 
-There needs a way to map the placement rules in SQL to PD placement rule configuration. Most of the fields are discussed above, so this part focuses on `group_id`, `id`, `start_key` and `end_key`.
+However, the definition is not valid. The unique index on `username` can not be enforced, because there are no global indexes for partitioned tables.
 
-`group_id` is used to identify the source of the placement rules, so `group_id` is `tidb`.
+In the future we will need to be able to define the index for `username` as a global index, and allow it to have different placement rules where it can be read from all regions.
 
-`ALTER PLACEMENT POLICY` and `DROP PLACEMENT POLICY` need to find the rules of a specified object efficiently. It can be achieved by encoding the object ID in `id`.
-
-However, an object may have multiple rules for a single role. For example:
+This example also demonstrates that this specification only provides partition level placement (and not row-level or column level security). The workaround for the user will require splitting the table:
 
 ```sql
-ALTER TABLE t
-	ALTER PLACEMENT POLICY CONSTRAINTS="{+zone=bj:2,+zone=sh:1}" ROLE=voter;
+CREATE TABLE users (
+	id INT NOT NULL auto_increment PRIMARY KEY,
+	/* public details */
+);
+
+CREATE TABLE user_details (
+	user_id INT NOT NULL,
+	/* private columns */
+	/* partition this table */
+);
 ```
 
-It needs 2 placement rules for `voter` in the PD placement rule configuration, because each rule can only specify one `count`. To make `id` unique, a unique identifier must be appended to `id`. DDL job ID plus an index in the job is a good choice.
+Assuming that global indexes can be added to the TiDB server, this use-case can be improved. But for correct execution the server will also require the following functionality:
+- The ability to mark global indexes as invisible at run-time if the `PLACEMENT POLICY` does not permit them to be read.
+- The ability to mark the clustered index as invisible. i.e. in the case that there is a global unique index on `username` it may be permitted to be read, but a read of the `username` from the clustered index might need to be disabled.
```

Comment:
any explaination about the second ability? why a global index with placement rule should turn back the access from cluster index.

Reply:
The global index here functions as a column-level access, so the placement policy allows it to be read by all regions, but the full row is not permitted to be read. This could also be implemented with "column level access" instead, where the clustered index might be able to be read, but certain columns are geo restricted.